titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/common/DifferenceResolverUtils.java

```diff
@@ -228,6 +232,50 @@ public static int countActiveNotStartedTasks(EntityHolder refJobHolder, EntityHo
         return pendingTaskIds.size();
     }
 
+    // Determines unassigned IP allocations based on the current reference view
+    public static Set<String> getUnassignedIpAllocations(JobView refJobView) {
+        // Get all IP allocations from the job
+        Set<String> unassignedIpAddressIds = refJobView.getJob().getJobDescriptor().getContainer().getContainerResources().getSignedIpAddressAllocations()
+                .stream()
+                .map(signedIpAddressAllocation -> signedIpAddressAllocation.getIpAddressAllocation().getAllocationId())
+                .collect(Collectors.toCollection(HashSet::new));
+
+        // Filter out those that are assigned
+        for (Task task : (List<Task>)refJobView.getTasks()) {
+            if (!TaskState.isTerminalState(task.getStatus().getState())) {
+                unassignedIpAddressIds.remove(task.getTaskContext().getOrDefault(TASK_ATTRIBUTES_IP_ALLOCATION_ID, ""));
+            }
+        }
+
+        return unassignedIpAddressIds;
+    }
+
+    // Copies specific task context entries, multiple if needed, from a previous task to a replacement task
+    private static Map<String, String> getTaskContextFromPreviousTask(Task previousTask) {
+        // Copy the IP allocation task context to the replacement task
+        String ipAllocationId = previousTask.getTaskContext().get(TASK_ATTRIBUTES_IP_ALLOCATION_ID);
+        return ipAllocationId != null ? Collections.singletonMap(TASK_ATTRIBUTES_IP_ALLOCATION_ID, ipAllocationId) : Collections.emptyMap();
+    }
+
+    // Sets the task context fields, multiple if needed, for an initial task.
+    private static Map<String, String> getInitialTaskContext(Set<String> unassignedIpAllocations) {
+        if (!unassignedIpAllocations.isEmpty()) {
+            String ipAllocationId = unassignedIpAllocations.iterator().next();
+            unassignedIpAllocations.remove(ipAllocationId);
+            return Collections.singletonMap(TASK_ATTRIBUTES_IP_ALLOCATION_ID, ipAllocationId);
+        }
+        return Collections.emptyMap();
+    }
+
+    // Get task context for a new or replacement task
```

Comment:
minor: javadoc style comments on `public` methods (`/** ... **/` instead of `// ...`)